File: 1805.number-of-different-integers-in-a-string.py
#
# @lc app=leetcode id=1805 lang=python3
#
# [1805] Number of Different Integers in a String
#

# @lc code=start
class Solution:
    def numDifferentIntegers(self, word: str) -> int:
        diff_num = set()
        digit_w = []
        for i in range(len(word)):
            if word[i].isdigit():
                digit_w.append(word[i])
            else:
                if digit_w != []:
                    for non_zero_ind in range(len(digit_w)):
                        if digit_w[non_zero_ind] != '0':
                            break
                    diff_num.add(''.join(digit_w[non_zero_ind:]))
                    digit_w.clear()
        if digit_w != []:
            for non_zero_ind in range(len(digit_w)):
                if digit_w[non_zero_ind] != '0':
                    break
            diff_num.add(''.join(digit_w[non_zero_ind:]))
            digit_w.clear()
        return len(diff_num)
# @lc code=end

File: cpp/1805.number-of-different-integers-in-a-string.cpp
/*
 * @lc app=leetcode id=1805 lang=cpp
 *
 * [1805] Number of Different Integers in a String
 */
#include <string>
#include <unordered_set>
using namespace std;
// @lc code=start
class Solution
{
public:
    int numDifferentIntegers(string word)
    {
        unordered_set<string> diff_num;
        int width = word.size();
        string digit_w;
        for (int i = 0; i != width; ++i)
        {
            if (isdigit(word[i]))
            {
                digit_w.push_back(word[i]);
            }
            else
            {
                if (digit_w != "")
                {
                    int non_zero_ind = digit_w.find_first_not_of("0");
                    if (non_zero_ind == -1)
                    {
                        diff_num.insert("0");
                    }
                    else
                    {
                        diff_num.insert(digit_w.substr(non_zero_ind));
                    }
                    digit_w.clear();
                }
            }
        }
        if (digit_w != "")
        {
            int non_zero_ind = digit_w.find_first_not_of("0");
            if (non_zero_ind == -1)
            {
                diff_num.insert("0");
            }
            else
            {
                diff_num.insert(digit_w.substr(non_zero_ind));
            }
        }
        return diff_num.size();
    }
};
// @lc code=end
